README
=head1 SYNOPSIS

    use List::Slice qw( head tail );

=head1 DESCRIPTION

This module provides functions for slicing lists. This is helpful when you
want to do a chain of manipulations on a list (map, grep, sort) and then
slice, without the cumbersome C<(...)[x]> syntax.

=head1 FUNCTIONS

=head2 head

    my @values = head $size, @list;

Returns the first C<$size> elements from C<@list>. If C<$size> is negative, returns
all but the last C<$size> elements from C<@list>.

    @result = head 2, qw( foo bar baz );
    # foo, bar

    @result = head -2, qw( foo bar baz );
    # foo

=head2 tail

    my @values = tail $size, @list;

Returns the last C<$size> elements from C<@list>. If C<$size> is negative, returns
all but the first C<$size> elements from C<@list>.

    @result = tail 2, qw( foo bar baz );
    # bar, baz

    @result = tail -2, qw( foo bar baz );
    # baz

=head1 SEE ALSO

L<List::Util>, L<List::MoreUtils>, L<List::UtilsBy>


CHANGES
Changelog for List-Slice

{{ $NAME }}

0.003 2015-05-31T22:28:48
 - add ppport.h to the root so non-dzil build works
 - fix segfault when negative tail is past the start of the list
 - ignore *.o files in the root dir

0.002 2015-05-24T04:35:40Z
 - add XS build artifacts to gitignore
 - remove Module::Build from the dist

0.001 2015-05-17T04:22:44Z
 - release v0.001
 - initial commit
 - Minting dist



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.036.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Slice-like operations on lists",
  "AUTHOR" => "Doug Bell <preaction\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "List-Slice",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "List::Slice",
  "PREREQ_PM" => {},
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


eg/bench.pl

use strict;
use warnings;
use List::Slice qw( head tail );
use List::Util qw( shuffle );
use Benchmark qw( cmpthese );

my @names = shuffle qw( aardvark albatross alligator alpaca ant anteater antelope ape
armadillo ass auk baboon badger barracuda bass bat bear beaver bee beetle bird
bison bittern bloodhound boar bobcat bovine buffalo bullfinch bullock butterfly
buzzard camel caribou cat caterpillar cattle cheetah chicken chimpanzee
chinchilla cicada clam cockroach cod colt coot cougar cow coyote crab crane
cricket crocodile crow deer dog dolphin donkey dotterel dove duck eagle eel
elephant elk ewe falcon ferret finch fish flamingo fly flying fish fowl fox
frog gerbil giraffe gnat gnu goat goldfinch goldfish goose gorilla goshawk
grasshopper greyhound grouse guinea pig gull hamster hare hawk hedgehog hen
heron herring hippopotamus hog hornet horse hound hummingbird hyena hyrax
impala jackrabbit jellyfish kangaroo kitten koala lark lemur leopard lice lion
llama lobster locust louse lynx mackerel magpie mallard manatee marten meerkat
mink minnow mole monkey moose mosquito mouse mule muskrat nighthawk nightingale
opossum orangutan ostrich otter owl ox oyster panda parrot partridge peacock
peafowl pelican penguin pheasant pig pigeon pike plover polar bear polecat pony
porcupine porpoise possum prairie dog quail rabbit raccoon racoon raptor rat
raven reindeer rhinoceros rook rooster salmon sandpiper sardine scorpion sea
horse sea lion seal shark sheep skunk snail snake snipe sow sparrow spider
squirrel stallion starling stork swallow swan swine swordfish Tasmanian devil
termite tiger toad tortoise toucan trout turkey turtle turtle dove viper
vulture wallaby walrus wasp weasel whale widgeon wild boar wolf wombat
woodchuck woodcock woodpecker worm wren yak zebra );

cmpthese -5, {
    perl => sub {
        my @found_names = grep { /^[a-d]/ } @names;
        my @topfive = @found_names[0..4];
    },
    slice => sub {
        my @topfive = ( grep { /^[a-d]/ } @names )[0..4];
    },
    head => sub {
        my @topfive = head 5, grep { /^[a-d]/ } @names;
    },
};

cmpthese -5, {
    perl => sub {
        my @found_names = grep { /^[a-d]/ } @names;
        my @bottomfive = @found_names[$#found_names-5..$#found_names];
    },
    head => sub {
        my @bottomfive = tail 5, grep { /^[a-d]/ } @names;
    },
};



lib/List/Slice.pm
package List::Slice;
# ABSTRACT: Slice-like operations on lists

use strict; use warnings;
require Exporter;

our @ISA        = qw(Exporter);
our @EXPORT_OK  = qw( head tail );

require XSLoader;
XSLoader::load('List::Slice', $List::Slice::VERSION);

1;

__END__

=head1 SYNOPSIS

    use List::Slice qw( head tail );

=head1 DESCRIPTION

This module provides functions for slicing lists. This is helpful when you
want to do a chain of manipulations on a list (map, grep, sort) and then
slice, without the cumbersome C<(...)[x]> syntax.

=cut

=head1 FUNCTIONS

=head2 head

    my @values = head $size, @list;

Returns the first C<$size> elements from C<@list>. If C<$size> is negative, returns
all but the last C<$size> elements from C<@list>.

    @result = head 2, qw( foo bar baz );
    # foo, bar

    @result = head -2, qw( foo bar baz );
    # foo

=head2 tail

    my @values = tail $size, @list;

Returns the last C<$size> elements from C<@list>. If C<$size> is negative, returns
all but the first C<$size> elements from C<@list>.

    @result = tail 2, qw( foo bar baz );
    # bar, baz

    @result = tail -2, qw( foo bar baz );
    # baz

=head1 SEE ALSO

L<List::Util>, L<List::MoreUtils>, L<List::UtilsBy>


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.036.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Slice-like operations on lists",
  "dist_author" => [
    "Doug Bell <preaction\@cpan.org>"
  ],
  "dist_name" => "List-Slice",
  "dist_version" => "0.001",
  "license" => "perl",
  "module_name" => "List::Slice",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "perl" => "5.006"
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.28",
  "Test::More" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
